core/drivers/crypto/crypto_api/acipher/rsassa.c

```diff
@@ -0,0 +1,918 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2020 NXP
+ *
+ * Brief   RSA Signature Software common implementation.
+ *         Functions preparing and/or verifying the signature
+ *         encoded string.
+ *
+ *         PKCS #1 v2.1: RSA Cryptography Standard
+ *         https://www.ietf.org/rfc/rfc3447.txt
+ */
+#include <crypto/crypto.h>
+#include <drvcrypt.h>
+#include <drvcrypt_asn1_oid.h>
+#include <drvcrypt_math.h>
+#include <malloc.h>
+#include <string.h>
+#include <tee_api_defines_extensions.h>
+#include <tee/tee_cryp_utl.h>
+#include <utee_defines.h>
+#include <util.h>
+
+#include "local.h"
+
+/*
+ * PKCS#1 V1.5 - Encode the message in Distinguished Encoding Rules
+ * (DER) format.
+ * Refer to EMSA-PKCS1-v1_5 chapter of the PKCS#1 v2.1
+ *
+ * @ssa_data  RSA data to encode
+ * @EM        [out] Encoded Message
+ */
+static TEE_Result emsa_pkcs1_v1_5_encode(struct drvcrypt_rsa_ssa *ssa_data,
+					 struct cryptobuf *EM)
+{
+	const struct drvcrypt_oid *hash_oid = NULL;
+	size_t ps_size = 0;
+	uint8_t *buf = NULL;
+
+	hash_oid =
+		&drvcrypt_hash_oid[TEE_ALG_GET_MAIN_ALG(ssa_data->hash_algo)];
```

Comment:
In practise we'll only get `TEE_MAIN_ALGO_MD5..TEE_MAIN_ALGO_SHA512` here, but `TEE_ALG_GET_MAIN_ALG()` can give larger values too for instance `TEE_MAIN_ALGO_SM2_PKE` (0x47).
Some check is needed to avoid reaching outside the bounds of `drvcrypt_hash_oid[]`. I wonder if it wouldn't be more safe to retrieve the hash OID via a function instead of indexing directly into an array.

Reply:
Is the hash algo not already checked in `TEE_Result rsassa_sign(struct drvcrypt_rsa_ssa *ssa_data)` ?

Reply:
You're right, it is checked. But it's still a bit tricky since it's a couple of function calls away and there's no way to assert here that the index isn't out of bounds. Anyone reviewing this function will jump each time they see this line, taking the focus from eventual more important things.

Reply:
Ok no problem, I will add a function to recover the hash instead

Reply:
Thanks

Reply:
Addressed in the last commit